app/models/user.rb

```diff
@@ -206,7 +207,42 @@ def self.authenticator(username = nil)
   end
 
   def self.authenticate(username, password, request = nil, options = {})
-    authenticator(username).authenticate(username, password, request, options)
+    user = authenticator(username).authenticate(username, password, request, options)
+    user.try(:link_user_to_session, request)
+
+    user
+  end
+
+  def link_user_to_session(request)
+    return unless request
+
+    session_id = request.session_options[:id]
+    sessions << Session.find_or_create_by(:session_id => session_id)
+  end
+
+  def broadcast_revoke_sessions
+    if Settings.server.session_store == "cache"
+      MiqQueue.broadcast(
+        :class_name  => self.class.name,
+        :instance_id => id,
+        :method_name => :revoke_sessions
+      )
+    else
+      # If using SQL or Memory, the sessions don't need to (or can't) be
+      # revoked via a broadcast since the session/token stores are not server
+      # specific, so execute it inline.
+      revoke_sessions
+    end
+  end
+
+  def revoke_sessions
+    current_sessions = Session.where(:user_id => id)
+    ManageIQ::Session.revoke(current_sessions.map(&:session_id))
+    current_sessions.destroy_all
+
+    TokenStore.token_caches.each do |_, token_store|
```

Comment:
`TokenStore.token_caches` can this be `nil` ?

Reply:
Nope:

https://github.com/ManageIQ/manageiq/blob/baafa9cca2be6a3aa77c7684528d3c88e260b9f4/lib/token_store.rb#L1-L2

Reply:
but in my console it returns nil. (without any previous api interaction)

it looks that it needs 

```ruby
def self.token_caches
  @token_caches
end
```

or 

`@@` for `token_caches` to become class variable.

Reply:
> @@ for token_caches to become class variable.

It should be a class variable in that context, since it is being assigned outside of a method (`@@` should be for when working within a method).

> but in my console it returns nil. (without any previous api interaction)

Ah, fair enough.  Perhaps the `cattr_reader` is messing with things.  I can just make this a regular method like `def self.token_caches` and use a `||=` to default assign it to `{}`.

Thanks for the heads up!

Reply:
> It should be a class variable in that context, since it is being assigned outside of a method (@@ should be for when working within a method).

You can use `@` at the class level...it's just a class-instance variable as opposed to a class variable.  However, if you want to access that from the class, then you need a class-level accessor as @lpichler described.  You don't need a class variable at all in this case.

Reply:
> Perhaps the cattr_reader is messing with things.

cattr_reader expects a class variable (`@@`)...if you want a class-instance variable (`@`), use `class_attribute :foo` or `class << self; attr_reader :foo; end`

Reply:
Huh, I guess I never really understood that there was that distinction between the two.  I just thought it was a different way of accessing the variable based on scope (`@@` for when you are in a `instance` context, or `@` when within a class context), but clearly I was wrong.

TIL.

Reply:
Anyway, probably just going to go with this:

```ruby
def self.token_caches
  @token_caches ||= {}
end
```

Which will be equivalent with what was there previously, but will allow it to have a method to access it from.

Reply:
it looks that it needs another adjustment as `@token_caches`  is not initialised in time when `acquire` is called.

```
TokenStore.acquire("api", 222)
=>
13: from bin/rails:4:in `<main>'
       12: from bin/rails:4:in `require'
       11: from railties (5.2.4.4) lib/rails/commands.rb:18:in `<top (required)>'
       10: from railties (5.2.4.4) lib/rails/command.rb:46:in `invoke'
        9: from railties (5.2.4.4) lib/rails/command/base.rb:69:in `perform'
        8: from thor (1.0.1) lib/thor.rb:392:in `dispatch'
        7: from thor (1.0.1) lib/thor/invocation.rb:127:in `invoke_command'
        6: from thor (1.0.1) lib/thor/command.rb:27:in `run'
        5: from railties (5.2.4.4) lib/rails/commands/console/console_command.rb:96:in `perform'
        4: from railties (5.2.4.4) lib/rails/commands/console/console_command.rb:19:in `start'
        3: from railties (5.2.4.4) lib/rails/commands/console/console_command.rb:64:in `start'
        2: from (irb):1
        1: from lib/token_store.rb:28:in `acquire'
NoMethodError (undefined method `[]' for nil:NilClass)
```

This works:

```
TokenStore.token_caches
TokenStore.acquire("api", 222)
=>
#<ActiveSupport::Cache::DalliStore:0x00007fa6cc29d248 @options={:namespace=>"MIQ:TOKENS:API", :threadsafe=>true, :expires_in=>222, :compress=>nil}, @data=#<Dalli::Client:0x00007fa6cc29d018 @servers=["127.0.0.1:11211"], @options={:namespace=>"MIQ:TOKENS:API", :threadsafe=>true, :expires_in=>222, :compress=>nil}, @ring=nil>>

```

and there is displayed issue in UI caused by obtaining token for `ws` (most likely related)

```
http://localhost:3090/api/auth?requester_type=ws
```

<img width="612" alt="Screenshot 2020-09-30 at 19 32 11" src="https://user-images.githubusercontent.com/14937244/94720031-38cc9580-0354-11eb-9daf-21a923ea2b3f.png">

```
 Error caught: [NoMethodError] undefined method `[]' for nil:NilClass
/Users/lpichler/projects/manageiq/manageiq/lib/token_store.rb:28:in `acquire'
/Users/lpichler/projects/manageiq/manageiq/lib/token_manager.rb:61:in `token_store'
/Users/lpichler/projects/manageiq/manageiq/lib/token_manager.rb:20:in `gen_token'
/Users/lpichler/projects/manageiq/manageiq/plugins/manageiq-api/lib/services/api/user_token_service.rb:40:in `generate_token'
/Users/lpichler/projects/manageiq/manageiq/plugins/manageiq-api/app/controllers/api/auth_controller.rb:8:in `show'
```

Reply:
Oh yeah, derp, good catch.  I will fix.

Reply:
This should be addressed now.